src/Raven.Server/Documents/Queries/QueryRunner.cs

```diff
@@ -92,7 +94,11 @@ public Task ExecutePatchQuery(string indexName, IndexQuery query, QueryOperation
             if (options.AllowStale == false && results.IsStale)
                 throw new InvalidOperationException("Cannot perform delete operation. Query is stale.");
 
-            foreach (var document in results.Results)
+            IEnumerable<Document> documents = results.Results;
+            if (options.MaxOpsPerSecond.HasValue)
+                documents = documents.LimitRate(options.MaxOpsPerSecond.Value, TimeSpan.FromSeconds(1));
+
+            foreach (var document in documents)
```

Comment:
Doesn't actually matter for our purposes.
You can't assume that you'll hit the batch size at the same time as the rate limit.
You processed a batch of 1024, and the batch size is 10000.
You processed batches very quickly, and you hit the limit. Now you have to wait a second while the timer goes down.
While you are holding the tx write lock and freezing whole db.